src/test/regress/input/temp_tablespaces.source

```diff
@@ -301,3 +301,106 @@ $$;
+ANALYZE tts_jazz;
+
+CREATE TABLESPACE temp_files LOCATION '@testtablespace@_temp_files';
+CREATE TABLESPACE temp_tables LOCATION '@testtablespace@_temp_tables';
+
+SET temp_file_tablespaces = temp_files;
+
+-- Should be in default tablespace.
+CREATE TEMPORARY TABLE tts_baz (a int) DISTRIBUTED BY (a);
+
+SELECT count(*)
+    FROM pg_class
+    WHERE pg_class.reltablespace = 0 AND relname = 'tts_baz';
+
+DROP TABLE tts_baz;
+RESET temp_file_tablespaces;
+
+CREATE OR REPLACE FUNCTION gp_tablespace_file_report(
+    temp_tablespaces text,
+    temp_file_tablespaces text
+) RETURNS TABLE (in_default boolean, in_temp_tables boolean, in_temp_files boolean) AS $$
+DECLARE
+    owo int;
+BEGIN
+    RESET temp_tablespaces;
+    RESET temp_file_tablespaces;
+
+    IF temp_tablespaces IS NOT NULL THEN
+        EXECUTE 'SET temp_tablespaces = ' || quote_literal(temp_tablespaces);
+    END IF;
+
+    IF temp_file_tablespaces IS NOT NULL THEN
+        EXECUTE 'SET temp_file_tablespaces = ' || quote_literal(temp_file_tablespaces);
+    END IF;
+
+    -- Create the temporary directory.
+    WITH out11 AS (
+        SELECT * FROM (
+            WITH cte(a,b) AS (SELECT a-1, b+1 FROM tts_foo)
+            SELECT * FROM (SELECT * FROM cte UNION ALL SELECT * FROM cte) AS X JOIN tts_bar ON b = c
+            UNION ALL SELECT *, 1, 2 FROM cte
+        ) AS XY
+        JOIN tts_jazz on c = e AND b = f
+    )
+    SELECT count(*) FROM out11 INTO owo;
+
+    -- Start monitoring temporary tablespace usage.
+    PERFORM gp_tablespace_watch_start(gp_execution_dbid(), 'default', gp_temptablespace_path((select dattablespace from pg_database where datname=current_database() limit 1)));
+    PERFORM gp_tablespace_watch_start(gp_execution_dbid(), 'temp_tables', gp_temptablespace_path((SELECT oid FROM pg_tablespace WHERE spcname = 'temp_tables' LIMIT 1)));
+    PERFORM gp_tablespace_watch_start(gp_execution_dbid(), 'temp_files', gp_temptablespace_path((SELECT oid FROM pg_tablespace WHERE spcname = 'temp_files' LIMIT 1)));
+
+    -- Run the query again to create temporary hashjoin files in the tablespaces.
+    WITH out11 AS (
+        SELECT * FROM (
+            WITH cte(a,b) AS (SELECT a-1, b+1 FROM tts_foo)
+            SELECT * FROM (SELECT * FROM cte UNION ALL SELECT * FROM cte) AS X JOIN tts_bar ON b = c
+            UNION ALL SELECT *, 1, 2 FROM cte
+        ) AS XY
+        JOIN tts_jazz on c = e AND b = f
+    )
+    SELECT count(*) FROM out11 INTO owo;
+
+    PERFORM gp_tablespace_watch_stop();
+
+    -- Check the existence of temporary hashjoin files.
+    RETURN QUERY SELECT
+        gp_tablespace_watch_match(gp_execution_dbid(), 'default', '_HashJoin') > 0,
+        gp_tablespace_watch_match(gp_execution_dbid(), 'temp_tables', '_HashJoin') > 0,
+        gp_tablespace_watch_match(gp_execution_dbid(), 'temp_files', '_HashJoin') > 0;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Files should be in default tablespace
+SELECT * FROM gp_tablespace_file_report(NULL, NULL);
```

Comment:
Where is test for global GUCs?
Where is test for several names in GUCs?
Why did not test check for table placement?